### PR TITLE
Java: Add the `XLEN` stream command

### PR DESCRIPTION
**Pull request by acarbonetto**
PR is dependent on the XTRIM command. 

https://valkey.io/commands/xlen/
https://redis.io/docs/latest/commands/xlen/

### CODE REVIEW

File: glide-core/src/protobuf/redis_request.proto [hunk 1]
```diff
@@ -149,6 +149,7 @@ enum RequestType {
     ZDiffStore = 106;
     SetRange = 107;
     ZRemRangeByLex = 108;
+    XLen = 109;
```
Comment (acarbonetto) [inline glide-core/src/protobuf/redis_request.proto:152]: TODO: probably needs an update

File: glide-core/src/request_type.rs [hunk 2]
```diff
@@ -117,6 +117,7 @@ pub enum RequestType {
     ZDiffStore = 106,
     SetRange = 107,
     ZRemRangeByLex = 108,
+    XLen = 109,
```
Comment (acarbonetto) [inline glide-core/src/request_type.rs:120]: TODO: ditto

Comment (Yury-Fridlyand) [reply glide-core/src/request_type.rs:120]: dritoo
pick at least 120 or 125

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 3]
```diff
@@ -1698,6 +1699,19 @@ public T xtrim(@NonNull String key, @NonNull StreamTrimOptions options) {
         return getThis();
     }
 
+    /**
+     * Returns the length, or number of entries in stream stored at <code>key</code>.
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1703]:
```suggestion
     * Returns the length, or number of entries in the stream stored at <code>key</code>.
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 4]
```diff
@@ -1595,17 +1595,7 @@ public void xadd_and_xtrim(BaseClient client) {
...
-            assertEquals(2L, ((RedisClient) client).customCommand(new String[] {"XLEN", key}).get());
-        } else if (client instanceof RedisClusterClient) {
-            assertEquals(
-                    2L,
-                    ((RedisClusterClient) client)
-                            .customCommand(new String[] {"XLEN", key})
-                            .get()
-                            .getSingleValue());
-        }
+        assertEquals(2L, client.xlen(key).get());
```
Comment (aaron-congo) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1598]: Should we update the test function name to indicate that it also tests xlen?

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 5]
```diff
@@ -1698,6 +1699,19 @@ public T xtrim(@NonNull String key, @NonNull StreamTrimOptions options) {
...
+     *
+     * @see <a href="https://valkey.io/commands/xlen/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @return The number of entries in the stream. If <code>key</code> does not exist, return <code>0
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:1707]:
```suggestion
     * @return Command Response - The number of entries in the stream. If <code>key</code> does not exist, return <code>0
```

File: java/integTest/src/test/java/glide/SharedCommandTests.java [hunk 6]
```diff
@@ -1636,40 +1616,23 @@ public void xadd_and_xtrim(BaseClient client) {
...
-                    ((RedisClusterClient) client)
-                            .customCommand(new String[] {"XLEN", key})
-                            .get()
-                            .getSingleValue());
-        }
+        assertEquals(1L, client.xlen(key).get());
 
         // Key does not exist - returns 0
-        assertEquals(0L, client.xtrim(key, new MaxLen(true, 1)).get());
+        assertEquals(0L, client.xtrim(key2, new MaxLen(true, 1)).get());
```
Comment (Yury-Fridlyand) [inline java/integTest/src/test/java/glide/SharedCommandTests.java:1626]: please update comment - stream is empty, key still exists
add a test where key really does not exist

**Review by Yury-Fridlyand — APPROVED** [review]

**Comment by Yury-Fridlyand** [comment]
Is this PR based on `xtrim`? Please add a note if so.

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 4 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1598]: sure

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 6 above)
Comment (acarbonetto) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1626]: what do you mean? 
> add a test where key really does not exist

Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java:1626]: Delete the stream/key

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 7]
```diff
@@ -73,4 +73,19 @@ public interface StreamBaseCommands {
...
+     * Returns the length, or number of entries in the stream stored at <code>key</code>.
+     *
+     * @see <a href="https://valkey.io/commands/xlen/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @return The number of entries in the stream. If <code>key</code> does not exist, return <code>0
+     *     </code>.
+     * @example
+     *     <pre>{@code
+     * Long num = client.xlen("key").get();
+     * assert num == 2L; // Stream at Key has 2 entries
```
Comment (aaron-congo) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:87]:
```suggestion
     * assert num == 2L; // Stream at "key" has 2 entries
```

**Review by aaron-congo — APPROVED** [review]

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java [hunk 8]
```diff
@@ -73,4 +73,19 @@ public interface StreamBaseCommands {
      * }</pre>
      */
     CompletableFuture<Long> xtrim(String key, StreamTrimOptions options);
+
+    /**
+     * Returns the length, or number of entries in the stream stored at <code>key</code>.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:78]: What is the difference between "length" and "number of entries"?

Comment (Yury-Fridlyand) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:78]: If it are same, please simplify the doc.

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 9]
```diff
@@ -2334,6 +2335,20 @@ public T xtrim(@NonNull String key, @NonNull StreamTrimOptions options) {
...
     }
 
+    /**
+     * Returns the length, or number of entries in the stream stored at <code>key</code>.
+     *
+     * @see <a href="https://valkey.io/commands/xlen/">valkey.io</a> for details.
+     * @param key The key of the stream.
+     * @return Command Response - The number of entries in the stream. If <code>key</code> does not
+     *     exist, return <code>0
+     *     </code>.
```
Comment (Yury-Fridlyand) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2345]: Easter egg from spotless

**Review by Yury-Fridlyand — APPROVED** [review]

File: java/client/src/main/java/glide/api/commands/StreamBaseCommands.java (hunk 8 above)
Comment (yipin-chen) [reply java/client/src/main/java/glide/api/commands/StreamBaseCommands.java:78]: Document uses "number of entries". So, we should remove "the length".

File: java/client/src/main/java/glide/api/models/BaseTransaction.java [hunk 10]
```diff
@@ -2334,6 +2335,20 @@ public T xtrim(@NonNull String key, @NonNull StreamTrimOptions options) {
         return getThis();
     }
 
+    /**
+     * Returns the length, or number of entries in the stream stored at <code>key</code>.
```
Comment (yipin-chen) [inline java/client/src/main/java/glide/api/models/BaseTransaction.java:2339]: Same as above, remove "the lenght"?

**Review by yipin-chen — COMMENTED** [review]
Minor comment update is required.

File: java/integTest/src/test/java/glide/SharedCommandTests.java (hunk 6 above)
Comment (Yury-Fridlyand) [reply java/integTest/src/test/java/glide/SharedCommandTests.java]: Try `del` and then `xlen` (on a missing key)

**Pull request closed without merging**